spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala

```diff
@@ -27,86 +27,153 @@ import org.apache.hadoop.fs.{Path, PathFilter}
-    import spark.implicits._
+   * Returns partitions that match the specified 'predicate'.
+   *
+   * @param spark a Spark session
+   * @param table a table name and (optional) database
+   * @param predicate a predicate on partition columns
+   * @return matching table's partitions
+   */
+  def getPartitionsByFilter(spark: SparkSession, table: String, predicate: String): Seq[SparkPartition] = {
+    val tableIdentifier = spark.sessionState.sqlParser.parseTableIdentifier(table)
+
+    val unresolvedPredicateExpr = spark.sessionState.sqlParser.parseExpression(predicate)
+    val resolver = spark.sessionState.analyzer.resolver
+    val plan = spark.table(table).queryExecution.analyzed
+    val resolvedPredicateExpr = unresolvedPredicateExpr.transform {
+      case attr: UnresolvedAttribute =>
+        plan.resolve(attr.nameParts, resolver) match {
+          case Some(resolvedAttr) => resolvedAttr
+          case None => throw new IllegalArgumentException(s"Could not resolve $attr using columns: ${plan.output}")
+        }
+    }
 
-    val expr = spark.sessionState.sqlParser.parseExpression(expression)
-    val partitions: Seq[(Map[String, String], Option[String], Option[String])] =
-      Hive.partitionsByFilter(spark, table, expr).map { p: CatalogTablePartition =>
-        (p.spec, p.storage.locationUri.map(String.valueOf(_)), p.storage.serde)
-      }
+    getPartitionsByFilter(spark, tableIdentifier, resolvedPredicateExpr)
+  }
+
+  /**
+   * Returns partitions that match the specified 'predicate'.
+   *
+   * @param spark a Spark session
+   * @param tableIdent a table identifier
+   * @param predicate a predicate on partition columns
+   * @return matching table's partitions
+   */
+  def getPartitionsByFilter(
+      spark: SparkSession,
+      tableIdent: TableIdentifier,
+      predicate: Expression): Seq[SparkPartition] = {
+
+    if (!predicate.resolved) {
+      throw new IllegalArgumentException(s"$predicate is not resolved")
+    }
 
-    partitions.toDF("partition", "uri", "format")
+    val catalog = spark.sessionState.catalog
+    val catalogTable = catalog.getTableMetadata(tableIdent)
+
+    catalog
+      .listPartitionsByFilter(tableIdent, Seq(predicate))
+      .map(catalogPartition => toSparkPartition(catalogPartition, catalogTable))
   }
 
   /**
    * Returns the data files in a partition by listing the partition location.
    *
-   * For Parquet partitions, this will read metrics from the file footer. For Avro partitions,
+   * For Parquet and ORC partitions, this will read metrics from the file footer. For Avro partitions,
+   * metrics are set to null.
+   *
+   * @param partition a partition
+   * @param conf a serializable Hadoop conf
+   * @param metricsConfig a metrics conf
+   * @return a Seq of [[SparkDataFile]]
+   */
+  def listPartition(
+      partition: SparkPartition,
+      conf: SerializableConfiguration,
+      metricsConfig: MetricsConfig): Seq[SparkDataFile] = {
```

Comment:
Why not default this MetricsConfig as well?

Reply:
If there are overloaded methods, Scala allows only one of them to have default args.